Add ingest scripts for s1 glofimr s1 imagery

moradology opened the pull request:
Add scripts to pull SAR imagery corresponding to the GLOFIMR dataset from sentinel-hub.

- [x] Provide adequate documentation to rerun process
- [x] Script generates STAC by pointing at correct S3 bucket/prefix
- [x] Reproject to 4326 🤦 

Closes #56

moradology committed:
Add ingest scripts for s1 glofimr s1 imagery

moradology committed:
Omit extremely lengthy example from function docs

moradology committed:
Deduplicate items

moradology committed:
Add documentation

moradology committed:
Document intended behavior of python code

moradology committed:
Add script to reproject tiffs on s3

moradology committed:
Organize dependencies and include requirements.txt

CloudNiner commented on catalogs/usfimr-s1/README.md:
```diff
@@ -0,0 +1,43 @@
+a user ID and a secret which jointly are sufficient to generate the token we will need to carry out
+Sentinel Hub API calls.
+
+Prepare the EU bucket following [these instructions](https://docs.sentinel-hub.com/api/latest/api/batch/#aws-s3-bucket-settings)
```
Could note here that a bucket is available in the noaa-flood AWS account already at

`s3://noaafloodmapping-sentinelhub-batch-eu-central-1`

CloudNiner commented on catalogs/usfimr-s1/build_catalog.py:
```diff
@@ -0,0 +1,145 @@
+
+    catalog = Catalog("glofimr-sar", catalog_description, title=catalog_title)
+
+    # We know the IDs used here, so we'll use them for keys
```
Man it sucks that GLOFIMR doesn't provide a consistent, useful id. 

Perhaps remove the TODO now that this is being merged and replace with an extra comment stating that this numeric id corresponds to the incrementing id in `ftp://guest:guest@sdml-server.ua.edu/USFIMR/USFIMR_all.zip` so future us can at least trace back to the actual events we're referring to here.

CloudNiner commented on catalogs/usfimr-s1/ingest_s1.py:
```diff
@@ -0,0 +1,165 @@
+    date_min = datetime.combine(flood_date, flood_start_time)
+    date_max = datetime.combine(flood_date, flood_end_time)
+
+    # Widen the min/max to hopefully get more results
```
Did this work? If not, probably just delete?

CloudNiner commented on catalogs/usfimr-s1/ingest_s1.py:
```diff
@@ -0,0 +1,165 @@
+                            flood.id
+                        )
+                    )
+                    print("SUCCESSFUL REQUEST ID: {}".format(batch_request_id))
```
Might be helpful to print the full s3 path that results were written to for easier parsing

CloudNiner commented on catalogs/usfimr-s1/ingest_s1_evalscript.js:
```diff
@@ -0,0 +1,27 @@
+//VERSION=3
```
Wooo you can write some js, read it into python script and it works.

I'd hate to have to maintain that multi-escaped single string nightmare it becomes.

CloudNiner commented on catalogs/usfimr-s1/request_builders.py:
```diff
@@ -0,0 +1,205 @@
+    return r
+
+
+def create_batch_request(flood_item, flood_temporal_bounds, connection_pool, token):
```
I realize we likely don't reuse this, but the method arguments seem overly specific for some of the methods in this file. 

In this particular case, perhaps swapping `flood_item` and `flood_temporal_bounds` to:
- `defaultTilePath`
- `geometry`
- `start_time`
- `end_time`

Would be both clearer for the person calling this method (i.e. its unclear what the type and shape of `flood_temporal_bounds` is without digging into the implementation) and easier to make generic later.

CloudNiner commented on catalogs/usfimr-s1/s3_stac_io.py:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python3
```
Huh. Surprised this doesn't exist in pystac already. I guess because of the boto3 dependency.

CloudNiner reviewed (commented):
Looks good overall to me. I didn't run it yet, because I wanted to get you the review so you have a chance to review comments before you leave for vacation. 

I'll run it tomorrow for a final test, fixup any issues I run into (hopefully zero!! 🤞) before merging. 

A couple of mostly take it or leave it suggestions for tweaks to docs/wording/code. 

I'm not totally sure how this will mesh with the no args `main.sh` ideas in #72 because you at least need to pass sentinel hub credentials -- perhaps they can be env vars without values in the docker-compose definition. If that's a blocker here on the rebase, just don't worry about it and I'll address it before merge.

CloudNiner commented on catalogs/usfimr-s1/build_catalog.py:
```diff
@@ -0,0 +1,145 @@
+
+    parsed_s3_path = urllib3.util.parse_url(args.imagery_root_s3)
+    bucket = parsed_s3_path.netloc
+    prefix = parsed_s3_path.path[1:]
```
Minor nitpick but this line fails with `TypeError: 'NoneType' object is not subscriptable`

If my imagery root is just a bucket: `--imagery-root-s3 s3://glofimr-sar-4326`

I can work around it with `--imagery-root-s3 s3://glofimr-sar-4326/` (add a slash to the end) but would be nice to not throw a cryptic error.

CloudNiner reviewed (changes requested):
Found a bug that needs addressing.

The asset Uris are incorrect for every instance after the first match to the sentinel hub chip id (e.g. `16SBE_3_0`).

In the screenshot below, I'm looking at the STAC item for ./catalogs/glofimr-s1/data/usfimr-sar-catalog/**3**/16SBE_3_0/16SBE_3_0.json but the asset Uris are all: s3://glofimr-sar-4326/**1**/654e5272-a61c-4d69-bc71-aef74ccaf85f/16SBE_3_0/

moradology committed:
Add docs; clarify some intent

moradology committed:
Make items unique across the global STAC

CloudNiner committed:
Refactor ingest_s1 to use requests lib

So that we don't need to explicitly include
the extra urllib3 dependency.

CloudNiner committed:
Use main.sh after rebase

- Parameterize sentinelhub output bucket
- Remove remaining usages of urllib3
- Add main.sh
- Add sentinelhub credential env vars to docker-compose definition

CloudNiner committed:
Update README

CloudNiner reviewed (approved):
Approving my own PR because @moradology is OoO.

Summary of changes today:
- Refactor to support `catalogs` docker container, which included parameterizing sentinel hub bucket name in ingest_s1.py and writing a main.sh that executes the three scripts of the pipeline in sequence
- Refactor to remove urllib3 as requests was already a dependency in `catalogs` container, and it reduces dependency tree and confusion
- Support installation of python deps via conda and pip in the `catalogs` docker container. Future users should only install via pip if an equivalent conda package is not available.
